### PR TITLE
Add torch._C.ScriptDict

### PR DESCRIPTION
**Pull request by SplitInfinity**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52659 [JIT] Add torch._C.ScriptDict**

**Summary**
This commit adds `torch._C.ScriptDict`, a dictionary type that has reference
semantics across the Python/TorchScript boundary. That is, modifications
made to instances of `torch._C.ScriptDict` in TorchScript are visible in
Python even when it is not returned from the function. Instances can be
constructed by passing an instance of a Python dictionary to
`torch.jit.script`. In the case of an empty dictionary, its type is
assumed to be `Dict[str, Tensor]` to be consistent with the handling of
empty dictionaries in TorchScript source code.

`torch.jit.dict` is implemented using a modified version of pybind's `stl_bind.h`-style bindings attached to `ScriptDict`, `ScriptDictIterator` and `ScriptDictKeyIterator`, wrapper classes around `c10::impl::GenericDict` and `c10::impl::GenericDict::iterator`. These bindings allow instances of `torch._C.ScriptDict` to be used as if it were a regular `dict` Python. Reference semantics are achieved by simply retrieving the `IValue` contained in `ScriptDict` in `toIValue` (invoked when converting Python arguments to `IValues` before calling TorchScript code).

**Test Plan**
This commit adds `TestScriptDict` to `test_list_dict.py`, a set of tests
that check that all of the common dictionary operations are supported
and that instances have reference semantics across the
Python/TorchScript boundary.

Differential Revision: [D27211605](https://our.internmc.facebook.com/intern/diff/D27211605)

### CODE REVIEW

**Comment by SplitInfinity** [comment]
@Lilyjjo 

It's not like *actually* ready for review in the sense that I need to polish some things up, add comments, etc. but I just wanted to bring this work to your attention.

File: torch/csrc/jit/python/pybind_utils.cpp [hunk 1]
```diff
@@ -136,6 +136,17 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
...
+      // If the object is a ScriptDict, retrieve the c10::Dict
+      // instance inside it.
+      try {
+        auto script_dict = py::cast<ScriptDict>(obj);
+        return script_dict.dict_;
+      } catch (py::cast_error& e) {
+      }
+
+      // If not (i.e. it is a regular Python dictionary), make a new
+      // c10::Dict.
```
Comment (SplitInfinity) [inline torch/csrc/jit/python/pybind_utils.cpp:149]: Add a warning telling users to script their dictionaries instead.

File: torch/csrc/jit/python/python_dict.cpp [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+          "__getitem__",
+          [](const std::shared_ptr<ScriptDict>& self, py::object key) {
+            try {
+              auto value = self->getItem(
+                  toIValue(std::move(key), self->type()->getKeyType()));
+              return toPyObject(value);
+            } catch (const std::out_of_range& e) { // Key doesn't exist.
+              throw py::key_error();
+            } catch (const py::cast_error& e) { // Key isn't the right type.
+              throw py::key_error();
```
Comment (SplitInfinity) [inline torch/csrc/jit/python/python_dict.cpp:130]: I just found out `TypeError` exists and that's what is thrown when you try to index a list with something that's not an integer. Perhaps that would be more appropriate here?

```suggestion
              throw py::type_error();
```

File: torch/csrc/jit/python/python_dict.cpp [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            return toPyObject(self->len());
+          })
+      .def(
+          "__contains__",
+          [](const std::shared_ptr<ScriptDict>& self, py::object key) {
+            try {
+              return toPyObject(self->contains(
+                  toIValue(std::move(key), self->type()->getKeyType())));
+            } catch (const py::cast_error& e) {
+              throw py::key_error();
```
Comment (SplitInfinity) [inline torch/csrc/jit/python/python_dict.cpp:117]:
```suggestion
              throw py::type_error();
```

File: torch/csrc/jit/python/python_dict.cpp [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+          [](const std::shared_ptr<ScriptDict>& self,
+             py::object key,
+             py::object value) {
+            IValue key_ivalue, value_ivalue;
+
+            // Try to convert the key to an IValue.
+            try {
+              key_ivalue = toIValue(std::move(key), self->type()->getKeyType());
+            } catch (const py::cast_error& e) {
+              throw py::key_error();
```
Comment (SplitInfinity) [inline torch/csrc/jit/python/python_dict.cpp:147]:
```suggestion
              throw py::type_error();
```

File: torch/csrc/jit/python/python_dict.cpp [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+            }
+
+            // Try to convert the value to an IValue.
+            try {
+              value_ivalue =
+                  toIValue(std::move(value), self->type()->getValueType());
+            } catch (const py::cast_error& e) {
+              throw py::value_error();
```
Comment (SplitInfinity) [inline torch/csrc/jit/python/python_dict.cpp:155]:
```suggestion
              throw py::type_error();
```

File: torch/csrc/jit/python/python_dict.cpp [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+      .def(
+          "__delitem__",
+          [](const std::shared_ptr<ScriptDict>& self, py::object key) {
+            IValue key_ivalue;
+
+            // Try to convert the key to an IValue.
+            try {
+              key_ivalue = toIValue(std::move(key), self->type()->getKeyType());
+            } catch (const py::cast_error& e) {
+              throw py::key_error();
```
Comment (SplitInfinity) [inline torch/csrc/jit/python/python_dict.cpp:169]:
```suggestion
              throw py::type_error();
```

File: torch/csrc/jit/python/pybind_utils.h [hunk 7]
```diff
@@ -696,13 +697,7 @@ inline py::object toPyObject(IValue ivalue) {
...
     return py::cast<py::object>(THPDevice_New(std::move(ivalue).toDevice()));
   } else if (ivalue.isGenericDict()) {
-    auto dict = std::move(ivalue).toGenericDict();
-    py::dict py_dict;
-    for (auto& pair : dict) {
-      py_dict[toPyObject(IValue{pair.key()})] =
-          toPyObject(IValue{pair.value()});
-    }
-    return std::move(py_dict);
+    return py::cast(ScriptDict(ivalue));
```
Comment (gmagogsfm) [inline torch/csrc/jit/python/pybind_utils.h:700]: Curious if this has caused any issues in internal CI.

File: torch/csrc/jit/python/python_dict.cpp [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+            ss << "Unable to infer type of dictionary: "
+               << inferred_type.reason();
+            throw JITException(ss.str());
+          }
+
+          type = inferred_type.type();
+        } else {
+          // If is empty, assume the type is Dict[str, Tensor] as is done in
+          // TorchScript code.
+          type = DictType::create(StringType::get(), TensorType::getInferred());
```
Comment (gmagogsfm) [inline torch/csrc/jit/python/python_dict.cpp:85]: @ansley Note that this is another place you might want to pay attention to when inferring type for `{}` and `[]`

File: torch/csrc/jit/python/python_dict.cpp (hunk 2 above)
Comment (gmagogsfm) [reply torch/csrc/jit/python/python_dict.cpp:130]: Yeah, `TypeError` seems more appropriate here. I think your test verifies that same exception type is thrown between eager and scripted dict, maybe we should have a test to cover/verify that?

Also, a minor suggestion, this try-catch-catch struct should be broken down into two try-catches. First one for key object to IValue and `TypeError` should be caught there, second one covers key lookup.

File: torch/csrc/jit/python/python_dict.cpp [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+              auto value = self->getItem(
+                  toIValue(std::move(key), self->type()->getKeyType()));
+              return toPyObject(value);
+            } catch (const std::out_of_range& e) { // Key doesn't exist.
+              throw py::key_error();
+            } catch (const py::cast_error& e) { // Key isn't the right type.
+              throw py::key_error();
+            }
+          },
+          py::return_value_policy::
```
Comment (gmagogsfm) [inline torch/csrc/jit/python/python_dict.cpp:133]: TIL

File: torch/csrc/jit/python/python_dict.cpp (hunk 4 above)
Comment (gmagogsfm) [reply torch/csrc/jit/python/python_dict.cpp:147]: ditto

File: torch/jit/_script.py [hunk 10]
```diff
@@ -783,11 +784,36 @@ def call_prepare_scriptable_func_impl(obj, memo):
...
+    Args:
+        obj (dict): The Python dictionary that is used to initialize the ``ScriptDict``
+                    returned by this function.
+
+    Returns:
+        An instance of ``torch._C.ScriptDict`` that has the same data as ``obj``
+        and can be passed between Python and TorchScript with reference semantics and
+        zero copy overhead.
+    """
+    # TODO: Does it make sense to pass in loc=None here?
```
Comment (gmagogsfm) [inline torch/jit/_script.py:808]: I think so. The only meaningful location to give is "current line", which is naturally available when error is thrown I think.

**Review by gmagogsfm — APPROVED** [review]
LGTM!

File: torch/csrc/jit/python/pybind_utils.h (hunk 7 above)
Comment (SplitInfinity) [reply torch/csrc/jit/python/pybind_utils.h:700]: It has. I didn't check them all, but I think they should be addressable.

File: torch/csrc/jit/python/python_dict.cpp (hunk 2 above)
Comment (SplitInfinity) [reply torch/csrc/jit/python/python_dict.cpp:130]: > Also, a minor suggestion, this try-catch-catch struct should be broken down into two try-catches. First one for key object to IValue and TypeError should be caught there, second one covers key lookup.

Why is that? Doesn't `try-catch` go through all the catch blocks in order until one of them "works"? Similar to `switch`?

Comment (gmagogsfm) [reply torch/csrc/jit/python/python_dict.cpp:130]: Oh, sorry for not being clear, it is a really minor concern. 

It seems that you want TypeError to catch exceptions rising from converting Python object for key to IValue, and you want `out_of_range`/`KeyError` to catch non-existent keys in dictionary. But, let's say there is a `out_of_range` rising from casting Python object key to IValue, then the exception would become a `KeyError` while it should really be `TypeError`.

Comment (SplitInfinity) [reply torch/csrc/jit/python/python_dict.cpp:130]: Yeah that's true

**Comment by SplitInfinity** [comment]
@gmagogsfm 

Re-requested review as discussed offline. The only thing that has changed is that `toPyObject` always returns a `py::dict`. This preserves reference semantics in the simplest case (i.e. user passes in a `ScriptDict` to a scripted function and does not return it) but breaks nested dictionary reference semantics. To reiterate, this simplification is to allow the foundational changes (e.g. `ScriptDict` class) to land without having to debug and fix every single fbcode failure.

File: torch/csrc/jit/python/python_dict.h [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    return dict_.toGenericDict().erase(key);
+  }
+
+  // Get the size of the dictionary.
+  int64_t len() const {
+    return dict_.toGenericDict().size();
+  }
+
+  // A c10::Dict instance that holds the actual data.
+  IValue dict_;
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/python/python_dict.h:124]: I tried this but was not able to default construct `GenericDict`. Let me try again.

File: torch/csrc/jit/python/python_dict.cpp [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+        return std::make_shared<ScriptDict>(data);
+      }))
+      .def(py::init([](py::dict dict, const TypePtr& type) {
+        auto data = toIValue(std::move(dict), type);
+        return std::make_shared<ScriptDict>(data);
+      }))
+      .def(
+          "__repr__",
+          [](const std::shared_ptr<ScriptDict>& self) {
+            return toPyObject(self->repr());
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/python/python_dict.cpp:95]: Hmm, I'm not sure if they are, I did it mostly out of habit. If pybind tries to convert them, that will be a no-op, right?

File: torch/jit/_script.py [hunk 13]
```diff
@@ -885,7 +888,30 @@ def _script_pdt(obj, optimize=None, _frames_up=0, _rcb=None, example_inputs: Opt
...
+        zero copy overhead.
+    """
+    if type_hint:
+        ty = ann_to_type(type_hint, None)
+        return torch._C.ScriptDict(obj, ty)  # type: ignore
+
+    return torch._C.ScriptDict(obj)  # type: ignore
+
+
+def script(obj, optimize=None, _frames_up=0, _rcb=None, type_hint=None):
```
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/jit/_script.py:914]: For empty dictionaries, I set the type to `Dict[str, Tensor]` if no type is passed in (consistent with the rest of TS). Why don't we just get rid of the this `type_hint` parameter?

File: torch/csrc/jit/python/python_dict.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/csrc/jit/python/python_dict.h]: Okay so it worked, but with the downside that I had to initialize the `dict_` as `dict_(AnyType::get(), AnyType::get())` because there is no default constructor for `GenericDict`. Let me know what you think!

File: torch/jit/_script.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (SplitInfinity) [reply torch/jit/_script.py]: I removed it as discussed offline.

File: torch/csrc/jit/python/python_dict.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply torch/csrc/jit/python/python_dict.cpp:91]: Should be

File: torch/jit/_script.py [hunk 14]
```diff
@@ -21,6 +21,7 @@
 from torch.utils import set_module
 from torch.jit._recursive import ScriptMethodStub, wrap_cpp_module, infer_methods_to_compile
 from torch.nn import Module
+from torch.jit.annotations import ann_to_type
```
Comment (gmagogsfm) [inline torch/jit/_script.py:24]: Minor lint error, please fix

**Review by gmagogsfm — APPROVED** [review]

**Comment by SplitInfinity** [comment]
@SplitInfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27211605).

**Comment by SplitInfinity** [comment]
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27211605).

**Comment by SplitInfinity** [comment]
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27211605).

**Comment by SplitInfinity** [comment]
@splitinfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27211605).

**Comment by SplitInfinity** [comment]
@SplitInfinity has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27211605).
